### PR TITLE
[WIP] Improvements to fieldset serialization format

### PR DESCRIPTION
**Pull request by jennybuckley**
Based on #102

Partially implements https://github.com/kubernetes/enhancements/pull/1336 (improvement 1+2)

Still missing:
- [x] String table differentiating character "!" at the start of each string table index in the fieldset
- [x] String table used in values, not just keys
- [x] String table index is written in base64
- [x] More complete string table
- [ ] Binary representation

### CODE REVIEW

**Comment by apelisse** [comment]
Can you also add a test to generate the set for the objects available in `internal/testdata`. This would provide a very simple way to compare v1 to v2.

File: internal/testdata/podset_v2.json [hunk 1]
```diff
@@ -0,0 +1 @@
+[2,0,!E8,0,!g,8,!DZ,[8,!Bk,[0,"app",0,"plugin1",0,"plugin2",0,"plugin3",0,"plugin4"],0,!i,0,!Ed,8,!I/,[11,{!R:"0a9d2b9e-779e-11e7-b422-42010a8001be"},[0,!E8,0,!Kc,0,!FC,0,!g,0,!i,0,!R]]],8,!r,[8,!FB,[11,{!i:"some-name"},[0,!V,8,!G,[11,{!i:"VAR_1"},[0,!i,8,!E1,[8,!HB,[0,!I,0,!i]]],11,{!i:"VAR_2"},[0,!i,8,!E1,[8,!HB,[0,!I,0,!i]]],11,{!i:"VAR_3"},[0,!i,8,!E1,[8,!HB,[0,!I,0,!i]]]],0,!5,0,!I4,0,!i,8,!El,[8,!Ds,[0,"cpu"]],0,!MD,0,!Mp,8,!HI,[11,{!Ec:"/var/run/secrets/kubernetes.io/serviceaccount"},[0,!Ec,0,!i,0,!Dp]]]],0,!EO,0,!Dd,0,!Dm,0,!Ht,0,!Hw,0,!JE,0,!Ia,0,!Kq,0,!Nb,0,!Gf,8,!C6,[11,{!i:"default-token-hu5jz"},[0,!i,8,!B0,[0,!GB,0,!Fd]]]],8,!B6,[8,!FA,[11,{!t:!Ih},[0,!Hg,0,!Ki,0,!B6,0,!t],11,{!t:"Initialized"},[0,!Hg,0,!Ki,0,!B6,0,!t],11,{!t:!Gl},[0,!Hg,0,!Ki,0,!B6,0,!t],11,{!t:"Ready"},[0,!Hg,0,!Ki,0,!B6,0,!t]],0,!J/,0,!Bi,0,!BD,0,!BE,0,!Do,0,!Es]]
```
Comment (lavalamp) [inline internal/testdata/podset_v2.json]: `/var/run/secrets/kubernetes.io/serviceaccount` should go in the string table, that is pretty common I think

File: fieldpath/strings/v1.go [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package strings
+
+var v1 = `
```
Comment (lavalamp) [inline fieldpath/strings/v1.go]: What is this?

File: fieldpath/strings/stream.go [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+var digits = []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/")
+
+func toBase64(x int) string {
+	if x == 0 {
+		return "A"
+	}
+	var s string
+	for x > 0 {
+		s = fmt.Sprintf("%v%v", string(digits[x%64]), s)
```
Comment (lavalamp) [inline fieldpath/strings/stream.go]: Writing these via `Write()` one-at-a-time might actually be better than FormatInt?

File: fieldpath/strings/reader.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+func parseBase64(b []byte) int {
+	n := 0
+	for _, d := range b {
+		if d >= byte('A') && d <= byte('Z') {
+			n = n*64 + int(d) - int('A')
+		} else if d >= byte('a') && d <= byte('z') {
+			n = n*64 + int(d) - int('a') + 26
+		} else if d >= byte('0') && d <= byte('9') {
+			n = n*64 + int(d) - int('0') + 52
```
Comment (lavalamp) [inline fieldpath/strings/reader.go]: hm I thought folks typically still used 0-9 to mean the same things as in decimal.

**Pull request closed without merging**